File: solutions/python3/0078.py
class Solution:
    def subsets(self, nums: List[int]) -> List[List[int]]:
        ans = []
        self.dfs(nums, len(nums), 0, [], ans)
        return ans

    def dfs(self, nums: List[int], target: int, s: int, curr: List[int], ans: List[List[int]]) -> None:
        if target < 0:
            return

        ans.append(curr)

        for i in range(s, len(nums)):
            self.dfs(nums, target - 1, i + 1, curr + [nums[i]], ans)

File: solutions/cpp/0078.cpp
class Solution {
public:
    vector<vector<int>> subsets(vector<int>& nums) {
        vector<vector<int>> ans;
        vector<int> curr;
        dfs(nums, nums.size(), 0, curr, ans);
        return ans;
    }

private:
    void dfs(vector<int>& nums, int target, int s, vector<int>& curr, vector<vector<int>>& ans) {
        if (target < 0) return;

        ans.push_back(curr);

        for (int i = s; i < nums.size(); i++) {
            curr.push_back(nums[i]);
            dfs(nums, target - 1, i + 1, curr, ans);
            curr.pop_back();
        }
    }
};